packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m

```diff
@@ -107,130 +94,180 @@ - (UIViewController *)viewControllerWithWindow:(UIWindow *)window {
 }
 
 /**
- * Returns the UIImagePickerControllerCameraDevice to use given [arguments].
- *
- * If the cameraDevice value that is fetched from arguments is 1 then returns
- * UIImagePickerControllerCameraDeviceFront. If the cameraDevice value that is fetched
- * from arguments is 0 then returns UIImagePickerControllerCameraDeviceRear.
+ * Returns the UIImagePickerControllerCameraDevice to use given [source].
  *
- * @param arguments that should be used to get cameraDevice value.
+ * @param source The source specification from Dart.
  */
-- (UIImagePickerControllerCameraDevice)getCameraDeviceFromArguments:(NSDictionary *)arguments {
-  NSInteger cameraDevice = [arguments[@"cameraDevice"] intValue];
-  return (cameraDevice == 1) ? UIImagePickerControllerCameraDeviceFront
-                             : UIImagePickerControllerCameraDeviceRear;
+- (UIImagePickerControllerCameraDevice)cameraDeviceForSource:(FLTSourceSpecification *)source {
+  switch (source.camera) {
+    case FLTSourceCameraFront:
+      return UIImagePickerControllerCameraDeviceFront;
+    case FLTSourceCameraRear:
+      return UIImagePickerControllerCameraDeviceRear;
+  }
 }
 
-- (void)pickImageWithPHPicker:(int)maxImagesAllowed API_AVAILABLE(ios(14)) {
+- (void)launchPHPickerWithContext:(nonnull FLTImagePickerMethodCallContext *)context
+    API_AVAILABLE(ios(14)) {
   PHPickerConfiguration *config =
       [[PHPickerConfiguration alloc] initWithPhotoLibrary:PHPhotoLibrary.sharedPhotoLibrary];
-  config.selectionLimit = maxImagesAllowed;  // Setting to zero allow us to pick unlimited photos
+  config.selectionLimit = context.maxImageCount;
   config.filter = [PHPickerFilter imagesFilter];
 
   _pickerViewController = [[PHPickerViewController alloc] initWithConfiguration:config];
   _pickerViewController.delegate = self;
   _pickerViewController.presentationController.delegate = self;
-
-  self.maxImagesAllowed = maxImagesAllowed;
+  self.callContext = context;
 
   [self checkPhotoAuthorizationForAccessLevel];
 }
 
-- (void)launchUIImagePickerWithSource:(int)imageSource {
+- (void)launchUIImagePickerWithSource:(nonnull FLTSourceSpecification *)source
+                              context:(nonnull FLTImagePickerMethodCallContext *)context {
   UIImagePickerController *imagePickerController = [self createImagePickerController];
   imagePickerController.modalPresentationStyle = UIModalPresentationCurrentContext;
   imagePickerController.delegate = self;
   imagePickerController.mediaTypes = @[ (NSString *)kUTTypeImage ];
+  self.callContext = context;
```

Comment:
Yes, that part is still a mess. This is a direct replacement for the existing ivar state, which had the same problem (just now consolidated to one object to make bookkeeping easier).

The reason I decided to explicitly make changing that not in scope (see PR description) is that I wanted to avoid turning the PR into a full rewrite. The pigeon changes are already substantial, but this keeps the actual logic flows mostly the same to limit the scope of what could be broken, and make the review more manageable.

We should absolutely fix that later.